### PR TITLE
NEFI-83 UI clients and Client provider and others

### PR DESCRIPTION
**Pull request by AlexanderPineda**
# Description

Client provider added together User Interfaces of client page, in contact management pages.

# Main Changes

- add custom component for modals in the project
- add table component for contact management pages
- client provider added to the project
- other formatting fixes and delete unused imports
- add the paginator custom component
- add validations for each form in the project

# NOTES
**Important: Please change the folder ***input*** to ***Input*** to the correct execution of the project**
**It is located in ***src/components/custom/Input*****

## Previews

imagen

- Real time validation proccess customized with react hook forms
imagen

- UI
imagen

imagen

- Client Provider added
imagen

- Validation for login page form
imagen

imagen

- paginator with customizable styles and buttons
imagen

### CODE REVIEW

File: client/src/components/forms/FormModalAddClient/index.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import Button from '../../custom/Button';
+import useStyles from './styles';
+import {FormControl} from '@mui/material';
+import {useTheme} from '@mui/material/styles';
+import {useForm} from 'react-hook-form';
+
+const FormModalAddClient = ({footerButtonText}) => {
+  const theme = useTheme();
+  const classes = useStyles(theme);
+  // const {handleSubmit} = useForm();
```
Comment (alphonse92) [inline client/src/components/forms/FormModalAddClient/index.js]: remove comments

File: client/src/hooks/clients/provider.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import PropTypes from 'prop-types';
+import context from './context';
+import {getClientsList} from '../../services/clients';
+
+export default function UserLoginProvider({children}) {
+  const [clients, setclients] = useState();
+
+  useLayoutEffect(() => {
+    getClients()
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (alphonse92) [inline client/src/hooks/clients/provider.js:11]: don't disable

File: client/src/hooks/user/Login/provider.js [hunk 3]
```diff
@@ -19,6 +19,7 @@ export default function UserLoginProvider({children}) {
     if (accessToken) {
       setUser(getUserFromLocalStorage());
     }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
```
Comment (alphonse92) [inline client/src/hooks/user/Login/provider.js:23]: dont disable

File: client/src/services/clients.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+export const getClientsList = async () => {
```
Comment (alphonse92) [inline client/src/services/clients.js:1]: That data will be paginated. Please use the conversations paginator features to know how to do that. And talk with camilo

**Review by alphonse92 — CHANGES_REQUESTED** [review]

File: client/src/components/TableContactManagement/index.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          <TableBody>
+            {clients && clients.map(row => (
+              <TableRow
+                className={`${classes.tableRow} ${classes.tableRowBody}`}
+                key={row.id + Math.random()}
+                sx={{'&:last-child td, &:last-child th': {border: '0'}}}>
+                <TableCell component="th" scope="row" align="center">
+                  {row.id}
+                </TableCell>
+                <TableCell align="center">{row.name}</TableCell>
```
[the remark this answers is not in the document]
Comment (AlexanderPineda) [reply client/src/components/TableContactManagement/index.js:42]: done

File: client/src/pages/dashboard/contactManagement/Clients/index.js [hunk 6]
```diff
@@ -1,48 +1,57 @@
...
-              <EmployeesInterface />
-            </div>
-          )) ||
-            (selectedTab === 1 && <h1>2</h1>)}
+        <div style={{width: '15%'}}>
+          <Modal
+            handleModal={handleModal}
+            modalTitle="Add Client"            
+            button={
+              <Button
```
[the remark this answers is not in the document]
Comment (AlexanderPineda) [reply client/src/pages/dashboard/contactManagement/Clients/index.js:37]: done

File: client/src/components/forms/FormModalAddClient/index.js (hunk 1 above)
Comment (AlexanderPineda) [reply client/src/components/forms/FormModalAddClient/index.js]: done

File: client/src/hooks/clients/provider.js (hunk 2 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/provider.js:11]: done

File: client/src/hooks/user/Login/provider.js (hunk 3 above)
Comment (AlexanderPineda) [reply client/src/hooks/user/Login/provider.js:23]: done

File: client/src/components/TableContactManagement/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexanderPineda) [reply client/src/components/TableContactManagement/index.js]: done

File: client/src/services/clients.js (hunk 4 above)
Comment (AlexanderPineda) [reply client/src/services/clients.js:1]: done

File: client/src/pages/dashboard/contactManagement/Clients/index.js [hunk 7]
```diff
@@ -1,48 +1,57 @@
...
         </div>
-        <div>
-          {(selectedTab === 0 && (
-            <div className={classes.employeesInterfaceContainer}>
-              <EmployeesInterface />
-            </div>
-          )) ||
-            (selectedTab === 1 && <h1>2</h1>)}
+        <div style={{width: '15%'}}>
+          <Modal
```
Comment (alphonse92) [inline client/src/pages/dashboard/contactManagement/Clients/index.js:33]: Move the line 33 to 45 to a new component, that is why I did mean with move out the implementation.

File: client/src/pages/dashboard/contactManagement/styles.js [hunk 8]
```diff
@@ -4,10 +4,11 @@ const useStyles = createUseStyles(
   {
     secondaryContainerLayout: {
       background: ({backgroundLight}) => backgroundLight,
-      padding: '40px',
+      padding: '40px 70px 40px 40px',
```
Comment (alphonse92) [inline client/src/pages/dashboard/contactManagement/styles.js:7]: No paddign without spacing. 70px = theme.spacing(9). theme.spacing(1)="8px" do the maths.

Comment (alphonse92) [reply client/src/pages/dashboard/contactManagement/styles.js:7]: 40px = theme.spacing(5)

File: client/src/hooks/clients/provider.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import React, {useState, useLayoutEffect, useCallback} from 'react';
+import PropTypes from 'prop-types';
+import context from './context';
+import {getClientsList} from '../../services/clients';
+import LoadClientsGeneratorHook from './LoadClientsGeneratorHook';
+
+export default function UserLoginProvider({children}) {
```
Comment (alphonse92) [inline client/src/hooks/clients/provider.js:7]: Rename this component

Comment (alphonse92) [reply client/src/hooks/clients/provider.js:7]: CustomerManagementProvider

File: client/src/components/custom/Paginator/index.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  return (
+    <Pagination
+      count={pages}
+      renderItem={item => (
+        <PaginationItem
+          components={{
+            last: props => <button {...props}>Last</button>,
+            next: props => <button {...props}>Next</button>,
+            first: props => <button {...props}>First</button>,
+            previous: props => <button {...props}>Previous</button>,
```
Comment (alphonse92) [inline client/src/components/custom/Paginator/index.js:20]: style the buttons. Variants text

File: client/src/pages/dashboard/contactManagement/Clients/styles.js [hunk 11]
```diff
@@ -2,38 +2,17 @@ import {createUseStyles} from 'react-jss';
...
+      marginTop: '-95px',
+      marginLeft: '100px',
+          
     },
+    button: {
+      height: '30px'      
+    }
   },
-  {name: 'contact-management'},
+  {name: 'clients'},
```
Comment (alphonse92) [inline client/src/pages/dashboard/contactManagement/Clients/styles.js:15]: contact-management-customers

File: client/src/hooks/clients/provider.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  const [clients, setclients] = useState();
+  const {nextPage} = LoadClientsGeneratorHook({page: 0, limit: 10});
```
Comment (alphonse92) [inline client/src/hooks/clients/provider.js:9]: LoadCustomersGeneratorHook

File: client/src/hooks/clients/LoadClientsGeneratorHook.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+import {API_URL} from '../../configs/config';
+import RequestGeneratorHook from '../RequestGeneratorHook';
+
+export default function LoadClientsGeneratorHook(...props) {
```
Comment (alphonse92) [inline client/src/hooks/clients/LoadClientsGeneratorHook.js:4]: Rename it to LoadCustomersGeneratorHook as well the module

**Review by alphonse92 — CHANGES_REQUESTED** [review]

File: client/src/pages/dashboard/contactManagement/Clients/index.js (hunk 7 above)
Comment (AlexanderPineda) [reply client/src/pages/dashboard/contactManagement/Clients/index.js:33]: done

File: client/src/hooks/clients/provider.js (hunk 9 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/provider.js:7]: done

File: client/src/pages/dashboard/contactManagement/Clients/styles.js (hunk 11 above)
Comment (AlexanderPineda) [reply client/src/pages/dashboard/contactManagement/Clients/styles.js:15]: done

File: client/src/hooks/clients/LoadClientsGeneratorHook.js (hunk 13 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/LoadClientsGeneratorHook.js:4]: done

File: client/src/components/custom/Paginator/index.js (hunk 10 above)
Comment (AlexanderPineda) [reply client/src/components/custom/Paginator/index.js:20]: done

File: client/src/components/ThreeDotsMenu/index.js [hunk 14]
```diff
@@ -7,7 +7,7 @@ import {DotsThreeOutlineVertical, DotsThreeOutline} from 'phosphor-react';
 import {useTheme} from '@mui/material/styles';
 import useStyles from './styles';
 
-export default function ThreeDotsMenu({orientation, menuItems}) {
+export default function ThreeDotsMenu({orientation, menuItems, menuLocation}) {
```
Comment (alphonse92) [inline client/src/components/ThreeDotsMenu/index.js:10]: Great

File: client/src/components/forms/FormModalAddClient/index.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+                {errors?.email?.message}
+              </FormHelperText>
+            )}
+          </FormControl>
+        </FormRowVertical>
+        <FormRowVertical>
+          <FormControl>
+            <label htmlFor="phoneFiled">Phone Number</label>
+            <Input
+              {...register('phone', {
```
Comment (alphonse92) [inline client/src/components/forms/FormModalAddClient/index.js:84]: Thanks for doing this job

File: client/src/components/TableContactManagement/styles.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+export default createUseStyles(
+  {
+    tableContainer:{
+      boxShadow: 'none'
+    },
+    tableRow: {      
+      borderBottom: theme => `7px solid ${theme?.variables?.backgroundLight}`,
+      '& td, & th': {
+        padding: '0px',
+        height: '40px'
```
Comment (alphonse92) [inline client/src/components/TableContactManagement/styles.js:12]: As I said: NOT fixed heights. Use paddings instead. Set a maxHeight if you want. Padding should be top: theme.spacing(2.5) and bottom: theme.spacing(2.5) + fontSize. maxHeight: theme.spacing(5)

File: client/src/components/TableContactManagement/styles.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    tableRowBody: {      
+      backgroundColor: theme => theme?.variables?.backgroundDark25,
+      '& td, & th': {
+        color: theme => theme?.variables?.backgroundDark
+      },
+    },
+    paginatorContainer: {
+      display: 'flex',
+      justifyContent: 'center',
+      paddingTop: '20px'
```
Comment (alphonse92) [inline client/src/components/TableContactManagement/styles.js:24]: paddingTop: theme.spacing(2.5)  (that is 20px)

File: client/src/components/forms/FormModalAddClient/styles.js [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    }),
+    footer: theme => ({
+      padding: theme.spacing(3),
+      display: 'flex',
+      width: '100%',
+      justifyContent: 'flex-end',
+      background: theme?.variables?.secondary,
+    }),
+    footerDivButton: {
+      width: '170px',
```
Comment (alphonse92) [inline client/src/components/forms/FormModalAddClient/styles.js:17]: Please remove that. Not fixed widths. If you need to set widths, please use it's parent container and use percentual widths.

File: client/src/hooks/clients/LoadCustomersGeneratorHook.js [hunk 19]
```diff
@@ -0,0 +1,7 @@
+import {API_URL} from '../../configs/config';
+import RequestGeneratorHook from '../RequestGeneratorHook';
+
+export default function LoadCustomersGeneratorHook(...props) {
+  const url = `${API_URL}/customers/`;
+  return RequestGeneratorHook(url, ...props);
```
Comment (alphonse92) [inline client/src/hooks/clients/LoadCustomersGeneratorHook.js:6]: You have to pull development and update with development. the second parameter of RequestGeneratorHook for this case is undefined

To update with development commit all your changes and :

```

git checkout development
git pull
git checkout $your_branch_here
git merge development

````


After that you should resolve conflicts. Resolve and commit the resolution

File: client/src/hooks/clients/use.js [hunk 20]
```diff
@@ -0,0 +1,11 @@
+import {useContext} from 'react';
+import currentContext from './context';
+
+export default function useCustomers() {
+  const context = useContext(currentContext);
+  const {clients, getNextPage} = context;
+  return {
+    clients,
```
Comment (alphonse92) [inline client/src/hooks/clients/use.js:8]: customers

File: client/src/hooks/clients/provider.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+      const pageResults = await nextPage();
+      const {results: data} = pageResults;
+      setclients(prev => [...prev, ...data]);
+    } catch (e) {
+      console.log(e);
+    }
+  }, [nextPage]);
+
+  const contextValue = {
+    clients,
```
Comment (alphonse92) [inline client/src/hooks/clients/provider.js:32]: customers

File: client/src/hooks/clients/provider.js [hunk 22]
```diff
@@ -0,0 +1,45 @@
+import React, {useState, useLayoutEffect, useCallback} from 'react';
+import PropTypes from 'prop-types';
+import context from './context';
+import {getClientsList} from '../../services/clients';
+import LoadCustomersGeneratorHook from './LoadCustomersGeneratorHook';
+
+export default function CustomerManagementProvider({children}) {
+  const [clients, setclients] = useState();
```
Comment (alphonse92) [inline client/src/hooks/clients/provider.js:8]: customers

**Review by alphonse92 — CHANGES_REQUESTED** [review]

File: client/src/components/ThreeDotsMenu/index.js (hunk 14 above)
Comment (AlexanderPineda) [reply client/src/components/ThreeDotsMenu/index.js:10]: check

File: client/src/components/TableContactManagement/styles.js (hunk 16 above)
Comment (AlexanderPineda) [reply client/src/components/TableContactManagement/styles.js:12]: maxHeight propperty does not have any effect in the row, the theme spacing has a similar effect comparing with the previous values with theme.spacing(0.3)

Comment (AlexanderPineda) [reply client/src/components/TableContactManagement/styles.js:12]: the height propperty was dropped according the requirement

File: client/src/components/TableContactManagement/styles.js (hunk 17 above)
Comment (AlexanderPineda) [reply client/src/components/TableContactManagement/styles.js:24]: already done

File: client/src/components/forms/FormModalAddClient/styles.js (hunk 18 above)
Comment (AlexanderPineda) [reply client/src/components/forms/FormModalAddClient/styles.js:17]: done

File: client/src/hooks/clients/LoadCustomersGeneratorHook.js (hunk 19 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/LoadCustomersGeneratorHook.js:6]: already merged

File: client/src/hooks/clients/use.js (hunk 20 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/use.js:8]: done

File: client/src/hooks/clients/provider.js (hunk 21 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/provider.js:32]: done

File: client/src/hooks/clients/provider.js (hunk 22 above)
Comment (AlexanderPineda) [reply client/src/hooks/clients/provider.js:8]: done

**Review by alphonse92 — APPROVED** [review]

**Pull request merged**